### ISSUE TITLE
How to build

### ISSUE THREAD

**Issue by narc-Ontakac2** [prrvchr/jdbcDriverOOo#12]
I am not finding a build system I know, actually I am not finding any build system.

**Comment by prrvchr** [comment]
> Documentation on how to build would be nice.

I admit there is nothing for now.

In fact everything is built in Eclipse IDE for Java Developers Version 2022-09 (4.25.0) with the extensions:
- [LOEclipse](https://libreoffice.github.io/loeclipse/)
- [PyDev](http://www.pydev.org/) Python IDE for Eclipse 12.1.0

**Comment by narc-Ontakac2** [comment]
Does this mean that there is no way to do a command line build?

I thought about building a Debian package from this (if it is not too much effort). This does however require a command line build.

**Comment by prrvchr** [comment]
> Does this mean that there is no way to do a command line build?

The only file to build from the source files is the Java archive jdbcDriverOOo.jar. I think LOEclipse uses a command line tool (but I don't know which one) to create this jar archive.

The jdbcDriverOOo.oxt file is just a compressed file and was produced by a bash script with the help of zip tool before any use of Eclipse.

**Comment by narc-Ontakac2** [comment]
> What are you really trying to build and what are the benefits that a Debian package can bring?

It can be configured to conflict with the 2 packages that need to be removed. In addition apt is to me "the" package manager. In this case packaging is however of minor importance. The libreoffice extensions are already packages. So I would only do packaging if is easy and this does not seem to be the case.

**Comment by prrvchr** [comment]
> It can be configured to conflict with the 2 packages that need to be removed.

Are you talking about both packages?
- libreoffice-sdbc-hsqldb
- libhsqldb1.8.0-java

If this is indeed the case, it is indeed urgent to replace these two packages which are almost 20 years old.

> So I would only do packaging if is easy and this does not seem to be the case.

If needed I can try to set up a build script.

**Comment by narc-Ontakac2** [comment]
> If this is indeed the case, it is indeed urgent to replace these two packages which are almost 20 years old.

Two questions on this:
1. Do you know why, lack of manpower?
2. Why didn't you modify the libreoffice source? If I recall the bug report someone asked you to do that.

**Comment by prrvchr** [comment]
> Do you know why, lack of manpower?

LibreOffice's concern has been to remove as much code as possible written in Java in order to make LibreOffice independent of Java. All this goes against a better integration of these two packages.

> Why didn't you modify the libreoffice source? If I recall the bug report someone asked you to do that.

No, the success of jdbcDriverOOo is to be written in Java. LibreOffice is written in C and then I was never asked to integrate Java code into LibreOffice code.

**Comment by narc-Ontakac2** [comment]
Thanks,I wasn't aware libreoffice is moving to C. This of course explains it.

**Comment by narc-Ontakac2** [comment]
I have give up on Debian packaging. The dependency on libslf4j-java can not be satisfied with a Debian package because that is on version 1.7.32 (Fun fact: apache-log4j is quite recent). 

Since Debian disallows including dependencies in a package (for security reasons, a vulnerable dependency can easily updated this way) I would have to update the libslf4j-java package. This exceeds my acceptable effort.

Since a documented way to build from source is something every open source project should have I leave the ticket open.

Thanks for the explanations.

**Comment by prrvchr** [comment]
> I would have to update the libslf4j-java package. This exceeds my acceptable effort.

I am not able to create a Debian package. However, we can always wait for this dependency to be updated to be able to make available this package which would replace the two obsolete packages.

> Since a documented way to build from source is something every open source project should have I leave the ticket open.

I will focus on writing a script capable of building from source files as Eclipse does with the LOEclipse extension, in order to make the build and use of Eclipse independent.

**Comment by narc-Ontakac2** [comment]
> I will focus on writing a script capable of building from source files as Eclipse does with the LOEclipse extension, in order to make the build and use of Eclipse independent.

I would recommend using one of the standard build systems (ant, gradle, maven). They are all in some way annoying, but documented and known to a lot of developers. If you are lucky Eclipse can create such a build script.

**Comment by prrvchr** [comment]
I have made some progress, it will soon be possible to build from Ant.

However, I have not completely understood the problem posed by the use of the slf4j-api-2.0.6.jar archive. Could you tell me more, in order to be able to resolve this last problem.

**Comment by narc-Ontakac2** [comment]
> However, I have not completely understood the problem posed by the use of the slf4j-api-2.0.6.jar archive. Could you tell me more, in order to be able to resolve this last problem.

While it is technically possible an official Debian package must not contain dependencies or static libraries. This makes security updates easily doable and avoids bloat.

**Comment by narc-Ontakac2** [comment]
I have submitted a [wishlst bug on the Debian package libslf4j-java](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1091916). For now I can package an inofficial package (provided no other obstruction occurs).

**Comment by prrvchr** [comment]
> I will replace those with dependencies.

You are right, all Python dependencies and their minimum version are listed in the `requirements.txt` file which is located at the root of the oxt file.

If python is not shipped with LibreOffice as is the case when installing LibreOffice using the package manager on Linux (at least with Ubuntu) then it is possible to separately provide the dependencies listed in the `requirements.txt` file and remove them from the `service/pythonpath/` directory of the oxt archive.

> After I managed to have a working package.

Glad to know this problem is moving forward, thank you...

**Comment by narc-Ontakac2** [comment]
There is (an ITP)[https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1107784] (Intent To Package bug) now. This communicates that I am on my way to package jdbcDriverOOo.

There are also several remarks and questions from my side:
1. There are a lot of files which are according to their headers are MIT licensed (e.g. ./source/jdbcDriverOOo/Addons.xcu).  The header takes precedence over the general statement in the README saying LGPL or MPL and I will list those separately. 
2. I can not package the jdbc drivers as part of the package. I will add them as "Suggests" dependencies. Could you point me to the location in the code where the list of JDBC drivers is appended to the UI. I will create a (Debian only) patch that only lists those that are installed.

**Comment by prrvchr** [comment]
> There are a lot of files which are according to their headers are MIT licensed (e.g. ./source/jdbcDriverOOo/Addons.xcu)

You're right, I hadn't realized that.
Normally the software is distributed under LGPL or MPL licenses. I need to correct that.

> I can not package the jdbc drivers as part of the package...

All drivers are located in the /driver folder.
They generally come from their respective websites, except for SQLite and Trino, where the archive is specific.

**Comment by narc-Ontakac2** [comment]
> You're right, I hadn't realized that. Normally the software is distributed under LGPL or MPL licenses. I need to correct that.

Thanks. This is not required but will make packaging easier.

> All drivers are located in the /driver folder. They generally come from their respective websites, except for SQLite and Trino, where the archive is specific.

The drivers will not be in the Debian package but will be installed separately. So I need to implement a little patch that only lists those that are actually installed on the system.  For that knowing the source code location where this list is generated would be helpful.

**Comment by prrvchr** [comment]
> For that knowing the source code location where this list is generated would be helpful.

I don't really have a file that lists all the drivers supported by jdbcDriverOOo.
Currently, they are located in the `source/jdbcDriverOOo/drivers` folder and also referenced in the LibreOffice XML configuration file `source/jdbcDriverOOo/Drivers.xcu`.

However, if necessary, I can maintain a Java dependency file for these drivers. I'm not sure which format (i.e. pom.xml) would be best.
